### DISCUSSION TITLE
String translations in a static context

### DISCUSSION THREAD

**Discussion by seyfeb** [perpetualKid/GetText.NET#54] [General]
If I use enumeration classes instead of simple enums, like presented, e.g., [here](https://learn.microsoft.com/en-us/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/enumeration-classes-over-enum-types), how would I use `GetText.NET` to use translatable names. See the example below, the strings would need to be made translatable in a static context.

```
public class Vegetable
    : Enumeration
{
    public static Vegetable Cucumber = new(1, "Cucumber");
    public static Vegetable BrownRice = new(2, "Brown Rice");

    public Vegetable(int id, string name)
        : base(id, name)
    {
    }
}
```

So when I use `Console.Write(Vegetable.BrownRice)` somewhere in my app, I want to get a translation of "Brown Rice" in the current culture.

**Comment by perpetualKid** [comment]
there is nothing specific with enumeration classes, and in a static context, you would just need to have a static catalog available

```csharp
    public class Vegetable : Enumeration
    {
        // This will load translations from "./locale/<CurrentUICulture>/Example.mo"
        private static ICatalog catalog = new Catalog("Example", "./locale");

        public static Vegetable Cucumber = new(1, "Cucumber");
        public static Vegetable BrownRice = new(2, catalog.GetString("Brown Rice"));

        public Vegetable(int id, string name)
            : base(id, name)
        {
        }
    }
```

**Reply by seyfeb** [reply]
I'm aware that this is a class like every other, I was just trying to provide context/a use case!

I don't think an enum should deal with the details of a catalog but I might try to find a way to inject a catalog or some service that handles the translation. The static context makes this a little bit more cumbersome.

At first, I was thinking of not touching the enumeration at all and only translate on usage

```
catalog.GetString(Vegetable.Cucumber);
```

but if I understand correctly, the extractor won't be able to check that there can only be those strings defined in the `Vegetable` class' static instances, even when I make the constructor private.

Would the extractor extract the strings if added as comments, i.e., like below?

```
// catalog.GetString("Cucumber")
public static Vegetable Cucumber = new(1, "Cucumber");
```

**Reply by perpetualKid** [reply]
if you want to use the comfort of Extractor, I'd suggest a private enum within your Vegetable class, redefining the Enumerable values to use descriptions

```csharp
    public class Vegetable : Enumeration
    {
        private enum Vegetables
        {
            [Description("Cucumber")]
            Cucumber,
            [Description("Brown Rice")]
            BrownRice,
        }

        public static Vegetable Cucumber = new(1, "Cucumber");
        public static Vegetable BrownRice = new(2, "Brown Rice");

        public Vegetable(int id, string name)
            : base(id, name)
        {
        }
    }


    internal class Program
    {
        private static Catalog catalog = new Catalog();
        private static Vegetable vegetable = Vegetable.BrownRice;
        static void Main(string[] args)
        {
            Console.WriteLine(catalog.GetString(vegetable.ToString()));
        }
    }

```
https://github.com/perpetualKid/GetText.NET/wiki/Enum-Descriptions
